sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -91,1012 +36,268 @@
-
-    /** Invoke {@link OldDoFn#processElement} on the bound {@code OldDoFn}. */
-    public void invokeProcessElement(
-        DoFn<InputT, OutputT>.ProcessContext c,
-        ExtraContextFactory<InputT, OutputT> extra);
-  }
-
-  /**
-   * Implementation of {@link DoFnReflector} for the arbitrary {@link DoFn}.
-   */
-  private static class GenericDoFnReflector extends DoFnReflector {
-
-    private final Method startBundle;
-    private final Method processElement;
-    private final Method finishBundle;
-    private final List<AdditionalParameter> processElementArgs;
-    private final List<AdditionalParameter> startBundleArgs;
-    private final List<AdditionalParameter> finishBundleArgs;
-    private final Constructor<?> constructor;
-
-    private GenericDoFnReflector(
-        @SuppressWarnings("rawtypes") Class<? extends DoFn> fn) {
-      // Locate the annotated methods
-      this.processElement = findAnnotatedMethod(ProcessElement.class, fn, true);
-      this.startBundle = findAnnotatedMethod(StartBundle.class, fn, false);
-      this.finishBundle = findAnnotatedMethod(FinishBundle.class, fn, false);
-
-      // Verify that their method arguments satisfy our conditions.
-      this.processElementArgs = verifyProcessMethodArguments(processElement);
-      this.startBundleArgs = verifyBundleMethodArguments(startBundle);
-      this.finishBundleArgs = verifyBundleMethodArguments(finishBundle);
-
-      this.constructor = createInvokerConstructor(fn);
-    }
-
-    private static Collection<Method> declaredMethodsWithAnnotation(
-        Class<? extends Annotation> anno,
-        Class<?> startClass, Class<?> stopClass) {
-      Collection<Method> matches = new ArrayList<>();
-
-      Class<?> clazz = startClass;
-      LinkedHashSet<Class<?>> interfaces = new LinkedHashSet<>();
-
-      // First, find all declared methods on the startClass and parents (up to stopClass)
-      while (clazz != null && !clazz.equals(stopClass)) {
-        for (Method method : clazz.getDeclaredMethods()) {
-          if (method.isAnnotationPresent(anno)) {
-            matches.add(method);
-          }
+      if (param.isSubtypeOf(BoundedWindow.class)) {
+        if (extraParameters.contains(DoFnSignature.ProcessElementMethod.Parameter.BOUNDED_WINDOW)) {
+          throw new IllegalArgumentException("Multiple BoundedWindow parameters");
         }
-
-        Collections.addAll(interfaces, clazz.getInterfaces());
-
-        clazz = clazz.getSuperclass();
-      }
-
-      // Now, iterate over all the discovered interfaces
-      for (Method method : ReflectHelpers.getClosureOfMethodsOnInterfaces(interfaces)) {
-        if (method.isAnnotationPresent(anno)) {
-          matches.add(method);
+        extraParameters.add(DoFnSignature.ProcessElementMethod.Parameter.BOUNDED_WINDOW);
+      } else if (param.isSubtypeOf(DoFn.InputProvider.class)) {
+        if (extraParameters.contains(DoFnSignature.ProcessElementMethod.Parameter.INPUT_PROVIDER)) {
```

Comment:
```
Preconditions.checkArgument(!extraParemeters.contains(...),
  "Multiple InputProvider parameters in %s", format(m));
```

So (1) use Preconditions.checkArgument (here and everywhere) and (2) make sure that the method is always described (eg., `format(m)`).